### ISSUE TITLE
Feature request to specify specific environment variables to inherit from calling process.

### ISSUE THREAD

**Issue by OmSaran** [apptainer/apptainer#1081]
Consider the following command: `mpiexec apptainer exec -e linux.sif bash -c "echo \$PMI_RANK"`. I would like the container to get the PMI_RANK environment variable _set by mpiexec_ (caller).  

This wouldn't produce the desired result since `PMI_RANK` is not inherited from mpiexec.

- Sure, one solution would be to remove `-e` but I want to selectively inherit only `PMI_RANK` from the caller. This wouldn't be possible to directly set in host via `--env` since value is determined at runtime by `mpiexec`. 
- Another solution I can think of is write a small wrapper script that calls apptainer (`apptainer exec -e --env PMI_RANK=$PMI_RANK ....`). 

It sure would be a lot more convenient to have a command line switch to do this in apptainer. Therefore I request for a new option in the cli, maybe like `--inherit-env PMI_RANK` to inherit the PMI_RANK only from the parent process. 

If this sounds like a reasonable feature request, I'd like to take a stab at implementing this by sending a pull request, thanks!

**Comment by GodloveD** [comment]
This is an interesting idea.  @anderbubble, as someone who has been investigating the limits of MPI with Apptainer recently, what do you think?

**Comment by anderbubble** [comment]
@OmSaran are there variables set by mpiexec that you do not want passed to the container? Or is your concern only variables from outside mpiexec that it would be passing on?

I'm thinking you generally would want any variables set by mpiexec; so I'd be more likely to try something like

```bash
env -i $(which mpiexec) apptainer exec linux.sif bash -c "echo \$PMI_RANK"
```

But if you do want to filter what's coming out of mpiexec (or if the issue is, perhaps, that some variables are being set by MPI's process launch process, e.g., during ssh) then I think this makes sense to support in Apptainer.

Personally, I'd prefer if we could just overload `--env` with this behavior, such that this would work, producing "myval":

```bash
env PMI_RANK=myval mpiexec apptainer exec -e linux.sif --env PMI_RANK bash -c "echo \$PMI_RANK"
```

But in testing we discovered that `--env`'s support for comma-separated values would make this a change in existing behavior.

```
env PMI_RANK=myval apptainer exec -e --env PMI_RANK=newval,OTHER_VAR alpine.sif sh -c "echo value of PMI_RANK: \$PMI_RANK, \$OTHER_VAR"
value of PMI_RANK: newval,OTHER_VAR,
```

So if the project would consider that too much of a breaking change from existing behavior, I think a new `--inherit-env` argument makes sense. Maybe the two could be merged back together in a later version.

**Comment by OmSaran** [comment]
@anderbubble 
> are there variables set by mpiexec that you do not want passed to the container?

Yes, I would need to have the option/flexibility to select what variables are passed on from `mpiexec` apart from discarding the non mpiexec host's environment variables.

> env PMI_RANK=myval mpiexec apptainer exec -e linux.sif --env PMI_RANK bash -c "echo \$PMI_RANK"

If mpiexec sets `PMI_RANK=newval` I assume it will echo `newval` and not `myval`. Please correct me if I'm wrong.

**Comment by anderbubble** [comment]
> If mpiexec sets PMI_RANK=newval I assume it will echo newval and not myval.

That would be my expectation as well.

**Comment by OmSaran** [comment]
>So if the project would consider that too much of a breaking change from existing behavior, I think a new --inherit-env argument makes sense. Maybe the two could be merged back together in a later version.

Who would make the call/decision on this?

**Comment by DrDaveD** [comment]
It would be up to the Technical Steering Committee members.  Probably the best way to decide would be to discuss it at community meeting.  Personally I don't yet understand the issue.
